src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
     end
-    listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
+
+    return listenloop(handler, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-function listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
-    try
-        id = 0
-        while isopen(tcpserver)
-            try
-                io = accept(tcpserver)
-                if !tcpisvalid(io; kw...)
-                    @info "Accept-Reject:  $io"
-                    close(io)
-                    continue
-                end
-                io = getsslcontext(io, sslconfig)
-                let i=id, conn = Connection(hostname, hostport, pipeline_limit, 0, require_ssl_verification, io)
-                    @async try
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"main server loop that accepts new tcp connections and spawns async threads to handle them"
+function listenloop(h, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
+    count = 1
+    while isopen(server)
+        try
+            io = accept(server)
+            io === nothing && continue
+            if !tcpisvalid(io)
+                verbose && @info "Accept-Reject:  $io"
+                close(io)
+                continue
+            end
+            connectioncounter[] += 1
+            conn = Connection(server.hostname, server.hostport, 0, 0, true, io)
+            let io=io, count=count
+                @async begin
+                    try
+                        verbose && @info "Accept ($count):  $conn"
+                        handle(h, conn, reuse_limit, readtimeout)
+                        verbose && @info "Closed ($count):  $conn"
                     catch e
-                        @error "Error ($i):  $conn" exception=(e, stacktrace(catch_backtrace()))
+                        @error exception=(e, stacktrace(catch_backtrace()))
                     finally
                         connectioncounter[] -= 1
-                        close(conn)
-                        @info "Closed ($i):  $conn"
+                        close(io)
+                        verbose && @info "Closed ($count):  $conn"
                     end
                 end
-            catch e
-                if e isa Base.IOError
-                    @warn "$e"
-                    break
-                else
-                    rethrow(e)
-                end
             end
-            id += 1
-        end
-    catch e
-        if e isa InterruptException
-            @warn "Interrupted: listen($hostname)"
-        else
-            rethrow(e)
```

Comment:
I can't say I'm completely happy with not rethrowing here.

I see that you're now logging the error and the stacktrace, but I'm pretty firmly of the belief that exceptions due unusual occurrences should be handled differently than exceptions due to bugs. i.e. contract violations (`ArgumentError`, `AssertionError`, `UndefVarError`) should almost never be caught and should almost always crash the program. There are two reasons for this: 1) bugs need to be fixed and crashing the program will be noticed and lead to a fix. 2)  Continuing to execute code aver a contract has been violated could result in undefined behaviour. If someone writes `@assert the data structure should never end up in this state` it is dangerous to keep running the program after that exception fails.

If you are a REPL, then you need to catch all the errors and show them to the user, but that is a special case. If you are a production server, then there should be some higher level supervisor (a cluster manager, or systemd or a custom health monitoring program) that notices that a this-should-never-occur error happened, and take appropriate action to restart the node/container/process and let someone know.

Consider the case where someone's Julia web server calls some module that calls some other module that keeps a global cache of important data. Maybe this code has a bunch of assertions in the code that updates the cache to check that mutating it doesn't end up in a bad state. But maybe the code that reads the cache just assumes the state is good, because consistency validation at read time would be too expensive. Now what if some call to the webserver does a data update that hits an `AssertionError`. If we don't crash the process, subsequent calls to read data will silently get invalid data.

An end user who knows "my app is not life and death" is free to put a `while try ... catch @error ... end` around their whole `main` function if they want it to run forever. But we should be very careful not to introduce places that where we suppress exceptions that the thrower intended to be fatal.

Perhaps you could have an optional layer in the `Hanlder/Router` stack that enables the user to elect to to catch fatal errors and has useful options for passing them to SNS, or a cluster manager, or some other notification service.